src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -2572,6 +2654,68 @@ address StubGenerator::generate_base64_decodeBlock() {
     __ BIND(L_bruteForce);
   }   // End of if(avx512_vbmi)
 
+  if (VM_Version::supports_avx2()) {
+    Label L_tailProc, L_topLoop, L_enterLoop;
+
+    // Check for buffer too small (for algorithm)
+    __ cmpl(length, 44);
+    __ jcc(Assembler::belowEqual, L_tailProc);
+
+    __ cmpl(isMIME, 0);
+    __ jcc(Assembler::notEqual, L_tailProc);
+
+    __ shll(isURL, 3);
+
+    // Set up constants
+    __ lea(r13, ExternalAddress(StubRoutines::x86::base64_AVX2_decode_tables_addr()));
+    __ vpbroadcastq(xmm4, Address(r13, isURL, Address::times_1), Assembler::AVX_256bit);  // 2F or 5F
+    __ vmovdqu(xmm10, xmm4);
+    __ vmovdqu(xmm12, Address(r13, 0x10));  // permute
+    __ vmovdqu(xmm9, Address(r13, 0x30));  // lut_hi
+    __ vmovdqu(xmm7, Address(r13, 0x50)); // merge
+    __ vmovdqu(xmm6, Address(r13, 0x70)); // merge mult
+    __ vmovdqu(xmm13, Address(r13, 0x90)); // shuffle
+
+    __ lea(r13, ExternalAddress(StubRoutines::x86::base64_AVX2_decode_URL_tables_addr()));
+    __ shll(isURL, 2);
+    __ vmovdqu(xmm11, Address(r13, isURL, Address::times_1, 0x00));  // lut_lo
+    __ vmovdqu(xmm8, Address(r13, isURL, Address::times_1, 0x20)); // lut_roll
+    __ shrl(isURL, 5);  // restore isURL
+    __ jmp(L_enterLoop);
+
+    __ align32();
+    __ bind(L_topLoop);
+    __ vpaddb(xmm0, xmm0, xmm2, Assembler::AVX_256bit);
+    __ vpmaddubsw(xmm0, xmm0, xmm7, Assembler::AVX_256bit);
+    __ vpmaddwd(xmm0, xmm0, xmm6, Assembler::AVX_256bit);
+    __ vpshufb(xmm0, xmm0, xmm13, Assembler::AVX_256bit);
+    __ vpermd(xmm0, xmm12, xmm0, Assembler::AVX_256bit);
+    __ subl(length, 0x20);
```

Comment:
Subtraction effects EFLAGs we can save one redundant compare per iteration on [#L2697](https://github.com/openjdk/jdk/pull/12126/files#diff-b938ab8a7bd9f57eb02271e2dd24a305bca30f06e9f8b028e18a139c4908ec92R2697)
by doing a prior subtraction by 0x2c (44) in pre-loop and increment by same amount post loop.

Same goes out for encode main loop also.